src/ucs/datastruct/ptr_array.h

```diff
@@ -120,13 +149,173 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
                     (UCS_V_INITIALIZED(_var), 0) : \
                     !__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index])))
 
+/**
+ * Check if element is free, and calculate how many free elements are ahead.
+ *
+ * @param [in]  _elem      An element in the ptr array.
+ * @param [out] _size_elem Pointer to the returned number of free elements ahead.
+ *
+ * @return 1 if the element is free and 0 if it's occupied.
+ */
+static inline int
+__ucs_ptr_array_is_free_and_get_step_size(ucs_ptr_array_elem_t _elem,
+                                          uint32_t *_size_elem)
+{
+    int _is_free = ((uintptr_t)(_elem) & UCS_PTR_ARRAY_FLAG_FREE);
+
+    if (_is_free) {
+       *_size_elem = _elem >> UCS_PTR_ARRAY_SIZE_FREE_SHIFT;
+    } else {
+       *_size_elem = 1;
+    }
+    ucs_assert(*_size_elem > 0);
+
+    return _is_free;
+}
 
 /**
  * Iterate over all valid elements in the array.
+ *
+ * @param [out] _var        Pointer to current array element in the foreach.
+ * @param [out] _index      Index variable to use as iterator (unsigned).
+ * @param [in]  _ptr_array  Pointer to a ptr array.
  */
 #define ucs_ptr_array_for_each(_var, _index, _ptr_array) \
-    for (_index = 0; _index < (_ptr_array)->size; ++_index) \
-         if (!__ucs_ptr_array_is_free(_var = (void*)((_ptr_array)->start[_index]))) \
+    uint32_t step_size = 1; \
+    for (_index = 0; (_index < (_ptr_array)->size); _index += step_size) \
+       if ((ucs_likely(!__ucs_ptr_array_is_free_and_get_step_size( \
+                            (ucs_ptr_array_elem_t)(_var = (void *)((_ptr_array)->start[_index])), \
+                            &step_size)))) \
+
 
+/*
+   Locked interface
+*/
+
+
+/* Locked ptr array */
+typedef struct ucs_ptr_array_locked {
+    ucs_ptr_array_t          super;
+    ucs_recursive_spinlock_t lock;
+} ucs_ptr_array_locked_t;
```

Comment:
@yosefe @brminich @alex--m @shuki-zanyovka 
what do you think about `ucs_ptr_array_safe_t` instead of `ucs_ptr_array_locked_t`?
and fix other places replacing `_locked_` by `_safe_` - we have other utilities that use `_safe_`

Reply:
`_safe_`  is not about the locking.
I'd discuss proposed use-case when locking is needed 
(my comment [here](https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fopenucx%2Fucx%2Fpull%2F4860%23discussion_r397081880&data=02%7C01%7Cmikhailb%40mellanox.com%7Ce4aff10f81b44536e7ad08d7cfe67d38%7Ca652971c7d2e4d9ba6a4d149256f461b%7C0%7C0%7C637206461157768914&sdata=tsHBCeq4uH3hvX%2FKVp9DlgEq7oR9XJDqCSeTzYhj4pw%3D&reserved=0))

Reply:
@brminich , @dmitrygx  - Please see my recommendation on: https://github.com/openucx/ucx/pull/4860#discussion_r397107821

Reply:
@brminich , @dmitrygx  - Please review the latest code.
Following a discussion with @alex--m , please see a nice and elegant solution for most concerns and comment.